eval/service.py

```diff
@@ -0,0 +1,695 @@
+from langchain_openai import ChatOpenAI
+
+from browser_use import Agent, Browser, BrowserConfig
+
+
+class Task:
+	def __init__(self, task_id, confirmed_task, website, reference_length, level):
+		self.task_id = task_id
+		self.confirmed_task = confirmed_task
+		self.website = website
+		self.reference_length = reference_length
+		self.level = level
+
+	def __str__(self):
+		return f'Task(task_id={self.task_id}, confirmed_task={self.confirmed_task}, website={self.website}, reference_length={self.reference_length}, level={self.level})'
+
+	def __repr__(self):
+		return self.__str__()
+
+
+class TaskTracker:
+	def __init__(self, task_id: str, task_text: str):
+		self.task_id = task_id
+		self.task_text = task_text
+		self.result_folder = Path(f'saved_trajectories/{task_id}')
+		self.trajectory_folder = self.result_folder / 'trajectory'
+		self.step_results = []
+		self.step_counter = 0
+		self.screenshots = []
+		self.setup_folders()
+
+	def setup_folders(self):
+		"""Create the necessary folder structure"""
+		self.result_folder.mkdir(parents=True, exist_ok=True)
+		self.trajectory_folder.mkdir(parents=True, exist_ok=True)
+
+	async def on_step_start(self, agent):
+		"""Record information at the start of a step"""
+		self.current_step = {'step_number': self.step_counter, 'start_time': datetime.now().isoformat(), 'actions': []}
+
+	async def on_step_end(self, agent):
+		"""Record information at the end of a step"""
+		# Take screenshot
+		browser_context = agent.browser_context
+		screenshot_b64 = await browser_context.take_screenshot()
+		screenshot_path = self.trajectory_folder / f'step_{self.step_counter}.png'
+
+		# Save screenshot to file
+		with open(screenshot_path, 'wb') as f:
+			f.write(base64.b64decode(screenshot_b64))
+
+		# Save screenshot path
+		self.screenshots.append(str(screenshot_path))
+
+		# Record action and result
+		if agent.state.last_result:
+			for result in agent.state.last_result:
+				self.current_step['actions'].append(
+					{
+						'content': result.extracted_content,
+						'error': result.error,
+						'is_done': result.is_done,
+						'success': result.success,
+					}
+				)
+
+		# Record end time
+		self.current_step['end_time'] = datetime.now().isoformat()
+		self.current_step['screenshot_path'] = str(screenshot_path)
+
+		# Add to step results
+		self.step_results.append(self.current_step)
+		self.step_counter += 1
+
+		# Save intermediate results
+		self.save_results()  # Save progress after each step
+
+	def save_results(self):
+		"""Save the consolidated results"""
+		# Create the final result object
+
+		formatted_result = {
+			'task_id': self.task_id,
+			'task': self.task_text,
+			'steps': self.step_results,
+			'action_history': [step['actions'][-1]['content'] for step in self.step_results],
+			'screenshot_paths': self.screenshots,
+			'final_result_response': self.step_results[-1]['actions'][-1]['content']
+			if self.step_results[-1]['actions'][-1]['is_done']
+			else None,
```

Comment:
personal preference but I kinda like the new python feature for stuff like this
```suggestion
			'final_result_response': (
				last_action['content']
				if (last_action := self.step_results[-1]['actions'][-1])['is_done']
				else None
			),
```